### ISSUE TITLE
Investigate atomic marker creation

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-eclipse#1634]
Check if we can use and profit from atomic marker creation in validation and generation
e.g.

https://git.eclipse.org/r/c/jdt/eclipse.jdt.core/+/176167/4/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/JavaBuilder.java

**Comment by ArneDeutsch** [comment]
I've done s syntetic test about how much time we could save. Obviously we can save time only in case we have many markers. If we have only few markers the creation process will be of no importance. Seeing many errors is not so common, but probably there are projects with many warnings and there it may be of importance indeed.

For my measurements I've created an Xtend file with 10.000 Errors (simple method calls to not existent method, 10.000 times). I then modified the code and saved. I instrumented the code to meaure the time the complete `ClusteringBuilderState.doUpdate` method has run and the time the `DefaultResourceUIValidatorExtension.createMarkers` method has run. I tried 5 times. Then I modified `MarkerCreator.createMarker` to use the new API and recorded the times again. Here is the raw data:

```
10000: 372ms
Complete: 2221ms
10000: 370ms
Complete: 1904ms
10000: 348ms
Complete: 2092ms
10000: 333ms
Complete: 1785ms
10000: 374ms
Complete: 2162ms

~360ms per run
~2030ms per run


10000: 214ms
Complete: 1643ms
10000: 195ms
Complete: 1743ms
10000: 188ms
Complete: 1523ms
10000: 190ms
Complete: 1856ms
10000: 184ms
Complete: 1504ms

~195ms per run
~1655ms per run
```

We can see, that we can expect to save around 30% - 50% of the time needed to create the markers resulting in a total save of 5%-10% (at least in this scenario). Of course this saving is only possible in case there are really that many markers.

**Comment by ArneDeutsch** [comment]
In case someone likes to run his own experiments ... here is a quick and dirty modification of MarkerCreator.createMarker that is using the new API:

```
	public void createMarker(Issue issue, IResource resource, String markerType) throws CoreException {
		Map<String, Object> attributes = new HashMap<>();
		
		String lineNR = "";
		if (issue.getLineNumber() != null) {
			lineNR = "line: " + issue.getLineNumber() + " ";
		}
		attributes.put(IMarker.LOCATION, lineNR + resource.getFullPath().toString());
		attributes.put(Issue.CODE_KEY, issue.getCode());		
		attributes.put(IMarker.SEVERITY, getSeverity(issue));
		attributes.put(IMarker.CHAR_START, issue.getOffset());
		if(issue.getOffset() != null && issue.getLength() != null)
			attributes.put(IMarker.CHAR_END, issue.getOffset()+issue.getLength());
		attributes.put(IMarker.LINE_NUMBER, issue.getLineNumber());
		attributes.put(Issue.COLUMN_KEY, issue.getColumn());
		attributes.put(IMarker.MESSAGE, issue.getMessage());

		if (issue.getUriToProblem()!=null) 
			attributes.put(Issue.URI_KEY, issue.getUriToProblem().toString());
		if(issue.getData() != null && issue.getData().length > 0) {
			attributes.put(Issue.DATA_KEY, Strings.pack(issue.getData()));
		}
		if (resolutionProvider != null && resolutionProvider.hasResolutionFor(issue.getCode())) {
			attributes.put(FIXABLE_KEY, true);
		}

		resource.createMarker(markerType, attributes);
	}

```

**Comment by cdietrich** [comment]
@ArneDeutsch did you also check DerivedResourceMarkers?

**Comment by ArneDeutsch** [comment]
No. actually I started with them, but was not be able to produce any markers here (tried with Xtend, but every error I introduced has lead to compilation issues, not in issues in Java source). But I think the results from above are transferable ... only in cases we have thousand or more markers that get be created the time will be noticable. In that cases we can save 30%-50% of the marker creation time. For each 10.000 Markers this will be 100-150ms (at least on my hardware).

**Comment by ArneDeutsch** [comment]
Looking at DerivedResourceMarkers again ... maybe I confused it with something else. Looking at the code I think there will be only 1 or maybe two markers per file. This will sum up very slowly. In a project with 10.000 of files critical time issues will be most probably at other places, not at marker creation time.

**Comment by ArneDeutsch** [comment]
To summarize:
The new API is much faster then the old one. We have only few places in our code (DerivedResourceMarkers, MarkerCreator and TaskMarkerCreator) that use this API. The needed changes are straight forward to implement and low risk. But not backwards compatible. We could be able to ship a fragment or an additional plugin that brings implementations of the three classes with new API usage. But it is unlikly that many users would really benefit (there projects need to have thouthands of errors/warnings/tasks but be very optimized in other ways and they must be on latest eclipse and Xtext verison).

I would suggest to implement the changes as soon as our minimal target platform allows but do NOT implement it now. Probably not worth the hassle. In projects that really have an issue with marker creation time a custom implementation of MarkerCreator (prototype above) could be done inside the project.

**Comment by cdietrich** [comment]
i assume we simply can use reflection too.

**Comment by ArneDeutsch** [comment]
I have run a test with reflective code as well. To my big surprise I could not measure any difference to the not reflective code. We could do that ....

**Comment by cdietrich** [comment]
@szarnekow what do you think?

**Comment by cdietrich** [comment]
see see the race from https://github.com/eclipse/xtext-eclipse/issues/1714#issue-962531806 too
